text/0000-partial_types.md

```diff
@@ -0,0 +1,642 @@
+- Feature Name: `partial_types`
+- Start Date: 2023-04-28
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+
+# Summary
+[summary]: #summary
+
+Partial Types proposal is a generalization on "partial borrowing"-like proposals.
+
+This proposal is universal flexible tool to work **safe** with partial parameters, partial arguments, partial references and partial borrowing.
+
+Advantages: maximum type safety, maximum type control guarantee, no ambiguities, flexibility, usability and universality.
+
+
+# Motivation
+[motivation]: #motivation
+
+Safe, Flexible controllable partial parameters for functions and partial consumption (including partial not borrowing) are highly needed.
+
+Partial Types extension gives to type-checker a **mathematical guarantee** that using _simultaneously_ partial typed variable, it multiple partial references and partial borrowing is as  **safe** as using them _at a sequence_. A guarantee to borrow-checker that by & var.{x,y} or &mut var.{z,t} borrowing the whole variable and pretending to borrow just several fields is **fully safe**.
+
+And since it is a guarantee by **type**, not by **values**, it has _zero cost_ in binary! Any type error is a compiler error, so no errors in the runtime.
+
+We could apply _theoretically_ this extension to all Product Types (`PT = T1 and T2 and T3  and ...`) and it fully covers all variants of possible uses of Product Types.
+
+So, most promised candidates are Structs and Tuples.
+
+This extension is not only fully backward-compatible, but is fully forward-compatible! Forward-compatibility is an ability to use updated functions old way.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+For Product Types `PT = T1 and T2 and T3  and ...` (structs, tuples) we assume they are like structs with controllable field-access.
+
+`<%permit>` - field is accessible (always implicit), `%deny` - field is forbidden to use, `%miss` - it is uninitialized field, field is forbidden to use.
+
+Let we wish to have next structure:
+```rust
+struct SR <T>{
+    val : T,
+    lnk : & T, // reference to val
+}
+```
+Now it is impossible to write immutable self-referential types. But with partial initialization it becomes easy:
+```rust
+let x = SR {val : 5i32 };
+    // omitted field is uninitialized, it has %miss field-access
+    // x : SR<i32>.{val, %miss lnk}
+
+x.lnk let= & x.val;
+    // (let=) "extender", late initialized field
+    // x : SR<i32>;
+    // x : SR<i32>.%full;
+```
+Easy-peasy! We create partial type and extend by late initializing fields!
+
+```rust
+let baz = (7i16, &51i32, "some_other_string").{1, 2, %unfill};
+    // or set access-filer before tuple constructor 
+    // baz : (i16, &i32, &str).{1, 2, %miss 0};
+
+let bar = (%miss 7i16, &51i32, "some_other_string");
```

Comment:
Agree, this sub-proposal is optional and it could be in "future possibilities".
It is almost useful for patterns only, when partial tuples are destructed.